src/v/kafka/server/handlers/fetch.cc

```diff
@@ -702,6 +711,186 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+class pid_controller {
+    // 0.1234 * scaling_factor = 1234
+    static constexpr int64_t scaling_factor = 10000;
+    static constexpr auto duration_unit = 1us;
+    static constexpr auto sample_rate = 250us;
+
+public:
+    pid_controller()
+      : _last_sampled_time{ss::sched_clock::now()}
+      , _p_coeff{config::shard_local_cfg().fetch_pid_p_coeff()}
+      , _i_coeff{config::shard_local_cfg().fetch_pid_i_coeff()}
+      , _d_coeff{config::shard_local_cfg().fetch_pid_d_coeff()}
+      , _target_fetch_sg_util{(
+          int64_t)(scaling_factor
+                   * config::shard_local_cfg().fetch_pid_target_fetch_sg())}
+      , _max_delay{config::shard_local_cfg().fetch_pid_max_debounce()}
+      , _p_coeff_binding{config::shard_local_cfg().fetch_pid_p_coeff.bind()}
+      , _i_coeff_binding{config::shard_local_cfg().fetch_pid_i_coeff.bind()}
+      , _d_coeff_binding{config::shard_local_cfg().fetch_pid_d_coeff.bind()}
+      , _target_fetch_sg_util_binding{config::shard_local_cfg()
+                                        .fetch_pid_target_fetch_sg.bind()}
+      , _max_delay_binding{
+          config::shard_local_cfg().fetch_pid_max_debounce.bind()} {
+        _p_coeff_binding.watch([this] {
+            reset();
+            _p_coeff = config::shard_local_cfg().fetch_pid_p_coeff();
+        });
+        _i_coeff_binding.watch([this] {
+            reset();
+            _i_coeff = config::shard_local_cfg().fetch_pid_i_coeff();
+        });
+        _d_coeff_binding.watch([this] {
+            reset();
+            _d_coeff = config::shard_local_cfg().fetch_pid_d_coeff();
+        });
+        _target_fetch_sg_util_binding.watch([this] {
+            reset();
+            _target_fetch_sg_util
+              = scaling_factor
+                * config::shard_local_cfg().fetch_pid_target_fetch_sg();
+        });
+        _max_delay_binding.watch([this] {
+            _max_delay = config::shard_local_cfg().fetch_pid_max_debounce();
+        });
+        setup_metrics();
+    }
+
+    std::chrono::milliseconds current_delay(const ss::scheduling_group& sg) {
```

Comment:
Is the whole scaling factor thing to replace floating point math with fixed-point integer math?

Reply:
Yeah, pretty much.

Reply:
Why? There's really no problem with using floating math in almost all cases. It's probably even going to be faster due to a lot of division (integer division is generally considerably slower than fp division), but I don't really think perf matters here since we are talking about a few ops even 250us or whatever.

Now we need to be careful of all the scaling gotchas like running out of precision, and also possible divide-by-zero for all the integer divisions (floating point div by zero just returns NaN, which will do "something" other than crash the server, in this case I guess it could affect the integral term which is carried over from period to period so I guess it bears consideration).

Reply:
Good point, switching back to floating point arithmetic.